tomcat-8.5/src/org/dogtagpki/tomcat/JSSContext.java

```diff
@@ -0,0 +1,102 @@
+package org.dogtagpki.tomcat;
+
+import java.util.List;
+import java.security.Provider;
+import java.security.Security;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+
+import org.apache.tomcat.util.net.SSLContext;
+
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.ssl.javax.JSSEngine;
+import org.mozilla.jss.ssl.javax.JSSParameters;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JSSContext implements org.apache.tomcat.util.net.SSLContext {
+    public static Logger logger = LoggerFactory.getLogger(JSSContext.class);
+
+    private javax.net.ssl.SSLContext ctx;
+    private JSSParameters params;
+
+    public JSSContext(String alias, List<String> negotiableProtocols) {
+        logger.debug("TomcatJSSContext(" + alias + ", ...");
+
+        params = new JSSParameters();
+        params.setAlias(alias);
+        params.setProtocols(negotiableProtocols.toArray(new String[0]));
+    }
+
+    public void init(javax.net.ssl.KeyManager[] kms, javax.net.ssl.TrustManager[] tms, java.security.SecureRandom sr) throws java.security.KeyManagementException {
+        logger.debug("TomcatJSSContext.init(...)");
+        try {
+            ctx = javax.net.ssl.SSLContext.getDefault();
+        } catch (java.security.NoSuchAlgorithmException e) {
+            throw new java.security.KeyManagementException(e.getMessage(), e);
+        }
+        ctx.init(kms, tms, sr);
+    }
+
+    public javax.net.ssl.SSLEngine createSSLEngine() {
+        logger.debug("TomcatJSSContext.createSSLEngine()");
+        JSSEngine eng = (JSSEngine) ctx.createSSLEngine();
+        eng.setSSLParameters(params);
+        return eng;
+    }
+
+    public javax.net.ssl.SSLSessionContext getServerSessionContext() {
+        logger.debug("TomcatJSSContext.getServerSessionContext()");
+        return ctx.getServerSessionContext();
+    }
+
+    public javax.net.ssl.SSLServerSocketFactory getServerSocketFactory() {
+        logger.debug("TomcatJSSContext.getServerSocketFactory()");
+        return ctx.getServerSocketFactory();
+    }
+
+    public javax.net.ssl.SSLParameters getSupportedSSLParameters() {
+        logger.debug("TomcatJSSContext.getSupportedSSLParameters()");
+        return params;
```

Comment:
AI @cipherboy: Figure out what this is supposed to return and how to implement it correctly.